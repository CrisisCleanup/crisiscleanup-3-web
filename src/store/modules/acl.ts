import Acl from 'browser-acl';
import type State from '@vuex-orm/core/dist/src/model/contracts/State';
import * as Sentry from '@sentry/browser';
import User from '../../models/User';

const AclState = {
  acl: new Acl(),
};

// Getters
const getters = {
  acl: (state: State) => state.acl,
};

// Actions
const actions = {};

// Mutations
const mutations = {
  addRule(state: State, { key, value }: Record<any, any>) {
    state.acl.addRule(key, value);
  },
  setUserAcl(state: State, user_id: any) {
    const user = User.find(user_id)!;
    const { permissions, beta_features } = user;
    Object.keys(permissions).forEach((permissionKey: string) => {
      state.acl.rule(permissionKey, user.permissions[permissionKey]);
    });
    for (const feature of beta_features) {
      state.acl.rule(`beta_feature.${feature}`, true);
    }

    state.acl.rule(
      `development_mode`,
      import.meta.env.NODE_ENV !== 'production',
    );
    state.acl.rule(`app_stage.${import.meta.env.VITE_APP_STAGE}`, true);

    Sentry.setUser({
      ...state.user,
      id: user.id,
      username: user.email,
      email: user.email,
    });
  },
};

export default {
  namespaced: true,
  state: AclState,
  getters,
  actions,
  mutations,
};
