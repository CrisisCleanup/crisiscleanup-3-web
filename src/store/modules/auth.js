import User from "@/models/User";
import { AuthService } from "@/services/auth.service";

const state = {
    user: AuthService.getUser()
};

// getters
const getters = {
    isLoggedIn: state => Boolean(state.user && state.user.access_token),
    userId: state => state.user ? state.user.user_claims.id : null
};

// actions
const actions = {
    async login ({ commit }, { email, password }) {
        let data = await User.api().login(email, password);
        commit('setUser', data.response.data);
        return data.response;
    },

    logout ({ commit }) {
        commit('setUser', null);
    }
};

// mutations
const mutations = {
    setUser (state, user) {
        state.user = user;
        if (!user) {
            AuthService.removeUser()
        } else {
            AuthService.saveUser(user);
        }
    }
};

export default {
    namespaced: true,
    state,
    getters,
    actions,
    mutations
}
