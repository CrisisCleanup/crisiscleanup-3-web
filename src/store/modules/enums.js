const AppState = {
  statuses: [],
  workTypes: [],
  locationTypes: [],
  phases: [],
};

// getters
const getters = {
  statuses: (state) => state.statuses,
  workTypes: (state) => state.workTypes,
  phases: (state) => state.phases,
  locationTypes: (state) => state.locationTypes,
  workTypeCommercialValues: (state) => {
    return Object.assign(
      {},
      ...state.workTypes.map((s) => ({ [s.key]: s.commercial_value })),
    );
  },
};

// actions
const actions = {};

// mutations
const mutations = {
  setStatuses(state, statuses) {
    state.statuses = statuses;
  },
  setWorkTypes(state, workTypes) {
    state.workTypes = workTypes;
  },
  setLocationTypes(state, locationTypes) {
    state.locationTypes = locationTypes;
  },
  setPhases(state, phases) {
    state.phases = phases;
  },
};

export default {
  namespaced: true,
  state: AppState,
  getters,
  actions,
  mutations,
};
