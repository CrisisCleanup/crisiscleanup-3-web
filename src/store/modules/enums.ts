import type State from '@vuex-orm/core/dist/src/model/contracts/State';

const AppState = {
  statuses: [],
  workTypes: [],
  locationTypes: [],
  phases: [],
  portal: null,
};

// Getters
const getters = {
  statuses: (state: State) => state.statuses,
  workTypes: (state: State) => state.workTypes,
  portal: (state: State) => state.portal,
  phases: (state: State) => state.phases,
  locationTypes: (state: State) => state.locationTypes,
  workTypeCommercialValues(state: State) {
    return Object.assign(
      {},
      ...state.workTypes.map((s: any) => ({ [s.key]: s.commercial_value })),
    );
  },
};

// Actions
const actions = {};

// Mutations
const mutations = {
  setStatuses(state: State, statuses: any[]) {
    state.statuses = statuses;
  },
  setWorkTypes(state: State, workTypes: any[]) {
    state.workTypes = workTypes;
  },
  setLocationTypes(state: State, locationTypes: any[]) {
    state.locationTypes = locationTypes;
  },
  setPhases(state: State, phases: any[]) {
    state.phases = phases;
  },
  setPortal(state: State, portal: any[]) {
    state.portal = portal;
  },
};

export default {
  namespaced: true,
  state: AppState,
  getters,
  actions,
  mutations,
};
