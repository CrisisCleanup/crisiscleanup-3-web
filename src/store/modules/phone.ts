import type State from '@vuex-orm/core/dist/src/model/contracts/State';

const AppState = {
  user: {},
  call: null,
  incomingCall: null,
  outgoingCall: null,
  stats: {},
  gateStats: [],
  agentStats: {},
  caller: null,
  gateway: {},
  needsWelcome: true,
  callState: 'AWAY',
  callType: null,
  languages: [],
  callHistory: [],
};

// Getters
const getters = {
  callState: (state: State) => state.callState,
  callType: (state: State) => state.callType,
  call: (state: State) => state.call,
  caller: (state: State) => state.caller,
  incomingCall: (state: State) => state.incomingCall,
  outgoingCall: (state: State) => state.outgoingCall,
  stats: (state: State) => state.stats,
  agentStats: (state: State) => state.agentStats,
  gateStats: (state: State) => state.gateStats,
  callHistory: (state: State) => state.callHistory,
  getUser: (state: State) => state.user,
  getGateway: (state: State) => state.gateway,
  isTakingCalls(state: State) {
    return state.callState !== 'AWAY';
  },
  isNotTakingCalls(state: State) {
    return !(state.callState !== 'AWAY');
  },
  isOnCall(state: State) {
    return (
      state.callState === 'ENGAGED-INBOUND' ||
      state.callState === 'ENGAGED-OUTBOUND'
    );
  },
  isInboundCall(state: State) {
    return state.callState === 'ENGAGED-INBOUND';
  },
  isOutboundCall(state: State) {
    return state.callState === 'ENGAGED-OUTBOUND';
  },
  isTransitioning(state: State) {
    return state.callState === 'TRANSITION';
  },
  getLanguages: (state: State) => state.languages,
};

// Actions
const actions = {};

// Mutations
const mutations = {
  setUser(state: State, user: any) {
    state.user = user;
  },
  setCurrentCall(state: State, call: any) {
    state.call = call;
  },
  setIncomingCall(state: State, call: any) {
    state.incomingCall = call;
  },
  setOutgoingCall(state: State, call: any) {
    state.outgoingCall = call;
  },
  setCaller(state: State, caller: any) {
    state.caller = caller;
  },
  setGateway(state: State, gateway: any) {
    state.gateway = gateway;
  },
  seenWelcome(state: State) {
    state.needsWelcome = false;
  },
  needsWelcome(state: State) {
    state.needsWelcome = true;
  },
  setState(state: State, newState: any) {
    state.callState = newState;
  },
  setCallType(state: State, newType: any) {
    state.callType = newType;
  },
  clearCall(state: State) {
    state.call = null;
    state.incomingCall = null;
    state.outgoingCall = null;
    state.caller = null;
    state.callType = null;
  },
  setGeneralStats(state: State, stats: any) {
    state.stats = stats;
  },
  setAgentStats(state: State, stats: any) {
    state.agentStats = stats;
  },
  setGateStats(state: State, stats: any) {
    state.gateStats = stats;
  },
  setLanguages(state: State, languages: any) {
    state.languages = languages;
  },
  setCallHistory(state: State, callHistory: any) {
    state.callHistory = callHistory;
  },
  resetState(state: State) {
    Object.assign(state, {
      user: {},
      call: null,
      incomingCall: null,
      outgoingCall: null,
      stats: {},
      agentStats: {},
      gateStats: [],
      caller: null,
      gateway: {},
      needsWelcome: true,
      callState: 'AWAY',
      callType: null,
      languages: [],
      callHistory: [],
    });
  },
};

export default {
  namespaced: true,
  state: AppState,
  getters,
  actions,
  mutations,
};
