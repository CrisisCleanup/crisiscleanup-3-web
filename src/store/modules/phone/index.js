/**
 * Phone store root
 */

import StreamsStore from '@/store/modules/phone/streams';
export default { StreamsStore };
