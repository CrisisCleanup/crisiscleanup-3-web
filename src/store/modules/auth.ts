import axios from 'axios';
import moment from 'moment';
import * as Sentry from '@sentry/vue';
import type State from '@vuex-orm/core/dist/src/model/contracts/State';
import { type ActionContext } from 'vuex';
import { AuthService } from '../../services/auth.service';
import type User from '../../models/User';

const AuthState = {
  user: AuthService.getUser(),
  showLoginModal: false,
};

// Getters
const getters = {
  isLoggedIn(state: State) {
    return state.user && AuthService.getAccessTokenExpiry().isAfter(moment());
  },
  isOrphan(state: State) {
    return state.user && !state.user.organization;
  },
  isOrganizationInactive(state: State) {
    return state.user && !state.user.organization?.is_active;
  },
  isAdmin(state: State) {
    return state.user && state.user.active_roles.includes(1);
  },
  userId: (state: State) => (state.user ? state.user.id : null),
  user: (state: State) => state.user,
  showLoginModal: (state: State) => state.showLoginModal,
};

// Actions
const actions = {
  async login(
    { commit }: ActionContext<never, never>,
    { email, password }: Record<string, string>,
  ) {
    const response = await axios.post(
      `${import.meta.env.VITE_APP_API_BASE_URL}/api-token-auth`,
      {
        email,
        password,
      },
    );
    commit('setUser', response.data);
    return response;
  },

  logout({ commit }: ActionContext<never, never>) {
    AuthService.logoutUser().then(() => {
      commit('setUser', null);
      const broadcast = new BroadcastChannel('logoutChannel');
      broadcast.postMessage('logout');
      window.location.href = '/';
    });
  },
};

// Mutations
const mutations = {
  setUser(state: State, user: User) {
    state.user = user;
    if (user) {
      AuthService.saveUser(user);
      Sentry.setUser({
        ...state.user,
        id: state.user.id,
        username: state.user.email,
        email: state.user.email,
      });
    } else {
      Sentry.setUser(null);
      AuthService.removeUser();
    }
  },
  setShowLoginModal(state: State, toggle: boolean) {
    state.showLoginModal = toggle;
  },
};

export default {
  namespaced: true,
  state: AuthState,
  getters,
  actions,
  mutations,
};
