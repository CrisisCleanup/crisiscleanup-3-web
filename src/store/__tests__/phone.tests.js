/**
 * phone.tests.js
 * Phone Store Tests
 */

import Agent from '@/models/Agent';
import { MockAgents, MockUsers } from '@/utils/testing';
import { actions } from '../modules/phone';

jest.mock('@/services/acs.service.js');
jest.mock('@/services/sso.service.js');

jest.mock('@/models/Worksite');
jest.mock('@/models/Incident');
jest.mock('@/models/User');
jest.mock('@/models/PhoneStatus');
jest.mock('@/models/Agent');

const MockMetrics = ({ queueCount = 2, online = 1 } = {}) => [
  {
    name: 'AGENTS_AVAILABLE',
    type: 'realtime',
    value: String(online),
  },
  {
    name: 'AGENTS_ONLINE',
    type: 'realtime',
    value: String(online),
  },
  {
    name: 'AGENTS_ON_CALL',
    type: 'realtime',
    value: '1',
  },
  {
    name: 'CONTACTS_IN_QUEUE',
    type: 'realtime',
    value: String(queueCount),
  },
  {
    name: 'CONTACTS_IN_QUEUE_OUTBOUND',
    type: 'realtime',
    value: '0',
  },
];

describe('metric actions', () => {
  it('getRealtimeMetrics', async () => {
    const commit = jest.fn();
    const state = {};

    const metricsIn = MockMetrics();
    await actions.getRealtimeMetrics({ commit, state }, { metrics: metricsIn });
    await actions.getRealtimeMetrics(
      { commit, state },
      { metrics: MockMetrics({ queueCount: 12 }) },
    );
    await actions.getRealtimeMetrics(
      { commit, state },
      { metrics: MockMetrics({ queueCount: 35, online: 0 }) },
    );
    await actions.getRealtimeMetrics(
      { commit, state },
      { metrics: MockMetrics({ queueCount: 35, online: 1 }) },
    );
    await actions.getRealtimeMetrics(
      { commit, state },
      { metrics: MockMetrics({ queueCount: -1, online: 1 }) },
    );

    expect(commit).toMatchSnapshot();
  });

  it('getAgentMetrics', async () => {
    const commit = jest.fn();
    const state = {
      controller: {
        history: {
          resolvedCases: [],
        },
      },
      contact: {
        attributes: {
          incidentId: 99,
        },
      },
    };

    const inboundMetrics = [
      {
        agent: 'xxxx',
        state: 'routable',
        user: MockUsers[0],
        total_calls: 6,
        total_inbound: 3,
        total_outbound: 3,
        total_rejects: 0,
        total_abandons: 0,
      },
      {
        agent: 'yyyy',
        state: 'routable',
        user: MockUsers[1],
        total_calls: 6,
        total_inbound: 3,
        total_outbound: 3,
        total_rejects: 0,
        total_abandons: 0,
        recent_contacts: [
          {
            caller_name: 'Adam Smith',
            dnis: 0,
            phone_number: '+11234567890',
            completed_at: '2019-12-31T06:00:00.000Z',
            cases: [1, 2],
            notes: 'metric note',
            status: 1,
          },
        ],
      },
    ];

    Agent.api = jest.fn(() => ({
      get: jest.fn(() => MockAgents[0]),
      getMetrics: jest.fn((id) =>
        inboundMetrics.find(({ agent }) => agent === id),
      ),
    }));

    const aniIncident = jest.fn();
    aniIncident.id = 99;

    await actions.getAgentMetrics(
      { commit, state, getters: { currentAniIncident: () => aniIncident } },
      {
        agents: [
          {
            agent_id: 'xxxx',
            state: 'routable',
            entered_timestamp: '2019-12-31T06:00:00.000Z',
          },
          {
            agent_id: 'yyyy',
            state: 'routable',
            entered_timestamp: '2019-12-31T06:00:00.000Z',
          },
        ],
      },
    );
    expect(commit.mock.calls).toMatchSnapshot();
  });
});
