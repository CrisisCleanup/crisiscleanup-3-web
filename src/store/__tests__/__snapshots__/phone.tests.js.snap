// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metric actions getAgentMetrics 1`] = `
Array [
  Array [
    "setControllerState",
    Object {
      "history": Object {
        "resolvedCases": Array [],
      },
    },
  ],
  Array [
    "setControllerState",
    Object {
      "history": Object {
        "resolvedCases": Array [
          1,
          2,
        ],
      },
    },
  ],
  Array [
    "setAgentMetrics",
    Object {
      "xxxx": Object {
        "agent": "xxxx",
        "currentState": "routable",
        "enteredTimestamp": 1577772000000,
        "recent_contacts": Array [],
        "state": "routable",
        "total_abandons": 0,
        "total_calls": 6,
        "total_inbound": 3,
        "total_outbound": 3,
        "total_rejects": 0,
        "user": Object {
          "first_name": "First",
          "id": 1,
          "last_name": "Last",
          "organization": Object {
            "id": 1,
            "name": "Test Organization",
          },
        },
      },
      "yyyy": Object {
        "agent": "yyyy",
        "currentState": "routable",
        "enteredTimestamp": 1577772000000,
        "recent_contacts": Array [
          Object {
            "caller_name": "Adam Smith",
            "cases": Array [
              Object {
                "case_number": "A196",
                "id": 1,
                "location": Object {
                  "coordinates": Array [
                    -82.0535959749408,
                    29.8670584991275,
                  ],
                  "type": "Point",
                },
                "name": "First Last",
                "updated_at": "2019-12-25T18:25:21.467128Z",
                "work_types": Array [
                  Object {
                    "claimed_by": null,
                    "status": "closed_incomplete",
                    "work_type": "flood_indoor",
                  },
                  Object {
                    "claimed_by": null,
                    "status": "closed_incomplete",
                    "work_type": "pda_1",
                  },
                ],
              },
              Object {
                "case_number": "A196",
                "id": 1,
                "location": Object {
                  "coordinates": Array [
                    -82.0535959749408,
                    29.8670584991275,
                  ],
                  "type": "Point",
                },
                "name": "First Last",
                "updated_at": "2019-12-25T18:25:21.467128Z",
                "work_types": Array [
                  Object {
                    "claimed_by": null,
                    "status": "closed_incomplete",
                    "work_type": "flood_indoor",
                  },
                  Object {
                    "claimed_by": null,
                    "status": "closed_incomplete",
                    "work_type": "pda_1",
                  },
                ],
              },
            ],
            "completed_at": "2019-12-31T06:00:00.000Z",
            "dnis": 0,
            "notes": "metric note",
            "phone_number": "+11234567890",
            "status": 1,
          },
        ],
        "state": "routable",
        "total_abandons": 0,
        "total_calls": 6,
        "total_inbound": 3,
        "total_outbound": 3,
        "total_rejects": 0,
        "user": Object {
          "first_name": "Test",
          "id": 2,
          "last_name": "Last",
          "organization": Object {
            "id": 1,
            "name": "Test Organization",
          },
        },
      },
    },
  ],
]
`;

exports[`metric actions getRealtimeMetrics 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "setMetrics",
      Object {
        "agentsAvailable": 1,
        "agentsNeeded": 0,
        "agentsOnCall": 1,
        "agentsOnline": 1,
        "contactsInQueue": 2,
        "contactsInQueueOutbound": 0,
        "totalWaiting": 2,
      },
    ],
    Array [
      "setMetrics",
      Object {
        "agentsAvailable": 1,
        "agentsNeeded": 0,
        "agentsOnCall": 1,
        "agentsOnline": 1,
        "contactsInQueue": 12,
        "contactsInQueueOutbound": 0,
        "totalWaiting": 12,
      },
    ],
    Array [
      "setMetrics",
      Object {
        "agentsAvailable": 0,
        "agentsNeeded": 3,
        "agentsOnCall": 1,
        "agentsOnline": 0,
        "contactsInQueue": 35,
        "contactsInQueueOutbound": 0,
        "totalWaiting": 35,
      },
    ],
    Array [
      "setMetrics",
      Object {
        "agentsAvailable": 1,
        "agentsNeeded": 2,
        "agentsOnCall": 1,
        "agentsOnline": 1,
        "contactsInQueue": 35,
        "contactsInQueueOutbound": 0,
        "totalWaiting": 35,
      },
    ],
    Array [
      "setMetrics",
      Object {
        "agentsAvailable": 1,
        "agentsNeeded": 0,
        "agentsOnCall": 1,
        "agentsOnline": 1,
        "contactsInQueue": 0,
        "contactsInQueueOutbound": 0,
        "totalWaiting": 0,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
